signer/Config.go

```diff
@@ -11,36 +16,231 @@ import (
-	CosignerThreshold int
-	ListenAddress     string
-	Nodes             []NodeConfig
-	Cosigners         []CosignerConfig
+func (cfg *CosignerConfig) LeaderElectMultiAddress() (string, error) {
+	addresses := make([]string, 1+len(cfg.Peers))
+	addresses[0] = cfg.P2PListen
+	for i, peer := range cfg.Peers {
+		addresses[i+1] = peer.P2PAddr
+	}
+	return client.MultiAddress(addresses)
+}
+
+type CosignerPeerConfig struct {
+	ShareID int    `json:"share-id" yaml:"share-id"`
+	P2PAddr string `json:"p2p-addr" yaml:"p2p-addr"`
 }
 
-func (cfg *Config) KeyFileExists() error {
-	if _, err := os.Stat(cfg.PrivValKeyFile); os.IsNotExist(err) {
-		return fmt.Errorf("private key share doesn't exist at path(%s)", cfg.PrivValKeyFile)
+func ValidateCosignerPeers(peers []CosignerPeerConfig, shares int) error {
+	// Check IDs to make sure none are duplicated
+	if dupl := duplicatePeers(peers); len(dupl) != 0 {
+		return fmt.Errorf("found duplicate share IDs in args: %v", dupl)
+	}
+
+	// Make sure that the peers' IDs match the number of shares.
+	for _, peer := range peers {
+		if peer.ShareID < 1 || peer.ShareID > shares {
+			return fmt.Errorf("peer ID %v in args is out of range, must be between 1 and %v",
+				peer.ShareID, shares)
+		}
+	}
+
+	// Check that exactly {num-shares}-1 peers are in the peer list, assuming
+	// the remaining peer ID is the ID the local node is configured with.
+	if len(peers) != shares-1 {
+		return fmt.Errorf("incorrect number of peers. expected (%d shares - local node = %d peers)",
+			shares, shares-1)
 	}
 	return nil
 }
 
+func duplicatePeers(peers []CosignerPeerConfig) (duplicates []CosignerPeerConfig) {
+	encountered := make(map[int]string)
+	for _, peer := range peers {
+		if _, found := encountered[peer.ShareID]; !found {
+			encountered[peer.ShareID] = peer.P2PAddr
+		} else {
+			duplicates = append(duplicates, CosignerPeerConfig{peer.ShareID, peer.P2PAddr})
+		}
+	}
+	return
+}
+
+func PeersFromFlag(peers string) (out []CosignerPeerConfig, err error) {
+	for _, p := range strings.Split(peers, ",") {
+		ps := strings.Split(p, "|")
+		if len(ps) != 2 {
+			return nil, fmt.Errorf("invalid peer string %s", p)
+		}
+		shareid, err := strconv.ParseInt(ps[1], 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, CosignerPeerConfig{ShareID: int(shareid), P2PAddr: ps[0]})
+	}
+	return
+}
+
+type ChainNode struct {
+	PrivValAddr string `json:"priv-val-addr" yaml:"priv-val-addr"`
+}
+
+func (cn ChainNode) Validate() error {
+	_, err := url.Parse(cn.PrivValAddr)
+	return err
+}
+
+type ChainNodes []ChainNode
+
+func (cns ChainNodes) Validate() error {
+	for _, cn := range cns {
+		if err := cn.Validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func ChainNodesFromArg(arg string) (out []ChainNode, err error) {
+	for _, n := range strings.Split(arg, ",") {
+		cn := ChainNode{PrivValAddr: n}
+		if err := cn.Validate(); err != nil {
+			return nil, err
+		}
+		out = append(out, cn)
+	}
```

Comment:
I think this should collect all the chain nodes and rely on `(ChainNodes).Validate`.